File: src/main/java/me/kuku/simbot/entity/GroupEntity.java
package me.kuku.simbot.entity;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "group_")
@NoArgsConstructor
@Getter
@Setter
@ToString(exclude = "qqEntities")
public class GroupEntity {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Integer id;
	@Column(unique = true, name = "group_")
	private Long group;
	private Boolean status = false;
	@ManyToMany(mappedBy = "groups")
	private Set<QqEntity> qqEntities = new HashSet<>();
	@Lob
	@Column(columnDefinition = "text")
	private String black;
	@Lob
	@Column(columnDefinition = "text")
	private String violation;
	@Lob
	@Column(columnDefinition = "text")
	private String qa;
	@Lob
	@Column(columnDefinition = "text")
	private String admin;
	@Lob
	@Column(columnDefinition = "text")
	private String weibo;
	@Lob
	@Column(columnDefinition = "text")
	private String biliBili;
	@Lob
	@Column(columnDefinition = "text")
	private String biliBiliLive;
	@Lob
	@Column(columnDefinition = "text")
	private String intercept;
	@Lob
	@Column(columnDefinition = "text")
	private String commandLimit;
	@Lob
	@Column(columnDefinition = "text")
	private String shellCommand;
	private Boolean colorPic = false;
	private Boolean recall = false;
	private Boolean pic = false;
	private Boolean leaveGroupBlack = false;
	private Boolean autoReview = false;
	private Boolean onTimeAlarm = false;
	private Integer maxCommandCountOnTime = -1;
	private Boolean locMonitor = false;
	private Boolean flashNotify = false;
	private Boolean repeat = true;
	private Boolean groupAdminAuth = false;
	private Boolean kickWithoutSpeaking = false;

	public GroupEntity(Long group){
		this.group = group;
	}

	public QqEntity getQq(Long qq){
		for (QqEntity qqEntity : qqEntities) {
			if (qqEntity.getQq().equals(qq)) return qqEntity;
		}
		return null;
	}

	public JSONArray getBlackJson(){
		return black == null ? new JSONArray() : JSON.parseArray(black);
	}

	public void setBlackJson(JSONArray json){
		this.black = json.toJSONString();
	}

	public JSONArray getViolationJson(){
		return violation == null ? new JSONArray() : JSON.parseArray(violation);
	}

	public void setViolationJson(JSONArray json){
		this.violation = json.toJSONString();
	}

	public JSONArray getQaJson(){
		return qa == null ? new JSONArray() : JSON.parseArray(qa);
	}

	public void setQaJson(JSONArray json){
		this.qa = json.toJSONString();
	}

	public JSONArray getAdminJson(){
		return admin == null ? new JSONArray() : JSON.parseArray(admin);
	}

	public void setAdminJson(JSONArray json){
		this.admin = json.toJSONString();
	}

	public JSONArray getWeiboJson(){
		return weibo == null ? new JSONArray() : JSON.parseArray(weibo);
	}

	public void setWeiboJson(JSONArray json){
		this.weibo = json.toJSONString();
	}

	public JSONArray getBiliBiliJson(){
		return biliBili == null ? new JSONArray() : JSON.parseArray(biliBili);
	}

	public void setBiliBiliJson(JSONArray json){
		this.biliBili = json.toJSONString();
	}

	public JSONArray getBiliBiliLiveJson(){
		return biliBiliLive == null ? new JSONArray() : JSON.parseArray(biliBiliLive);
	}

	public void setBiliBiliLiveJson(JSONArray json){
		this.biliBiliLive = json.toJSONString();
	}

	public JSONArray getInterceptJson(){
		return intercept == null ? new JSONArray() : JSON.parseArray(intercept);
	}

	public void setInterceptJson(JSONArray json){
		this.intercept = json.toJSONString();
	}

	public JSONObject getCommandLimitJson(){
		return commandLimit == null ? new JSONObject() : JSON.parseObject(commandLimit);
	}

	public void setCommandLimitJson(JSONObject json){
		this.commandLimit = json.toJSONString();
	}

	public JSONArray getShellCommandJson(){
		return shellCommand == null ? new JSONArray() : JSON.parseArray(shellCommand);
	}

	public void setShellCommandJson(JSONArray json){
		this.shellCommand = json.toJSONString();
	}
}

File: src/main/kotlin/me/kuku/simbot/entity/GroupEntity.kt
package me.kuku.simbot.entity

import com.alibaba.fastjson.JSON
import com.alibaba.fastjson.JSONArray
import com.alibaba.fastjson.JSONObject
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import javax.annotation.Resource
import javax.persistence.*

@Entity
@Table(name = "group_")
data class GroupEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private val id: Int? = null,
    @Column(unique = true, name = "group_")
    var group: Long = 0L,
    var status: Boolean? = false,
    @ManyToMany(mappedBy = "groups")
    var qqEntities: Set<QqEntity> = HashSet(),
    @Lob
    @Column(columnDefinition = "text")
    var black: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var violation: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var qa: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var admin: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var weibo: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var biliBili: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var biliBiliLive: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var intercept: String? = "[]",
    @Lob
    @Column(columnDefinition = "text")
    var commandLimit: String? = "{}",
    @Lob
    @Column(columnDefinition = "text")
    var shellCommand: String? = "[]",
    var colorPic: Boolean? = false,
    var recall: Boolean? = false,
    var pic: Boolean? = false,
    var leaveGroupBlack: Boolean? = false,
    var autoReview: Boolean? = false,
    var onTimeAlarm: Boolean? = false,
    var maxCommandCountOnTime: Int? = -1,
    var locMonitor: Boolean? = false,
    var flashNotify: Boolean? = false,
    var repeat: Boolean? = true,
    var groupAdminAuth: Boolean? = false,
    var kickWithoutSpeaking: Boolean? = false,
){
    companion object{
        fun getInstance(group: Long): GroupEntity{
            return GroupEntity(group = group)
        }
    }

    override fun toString(): String {
        return "";
    }

    override fun hashCode(): Int {
        return 0
    }

    fun getQq(qq: Long): QqEntity? {
        for (qqEntity in qqEntities) {
            if (qqEntity.qq == qq) return qqEntity
        }
        return null
    }

    var blackJson: JSONArray
        get() = if (black == null) JSONArray() else JSON.parseArray(black)
        set(json) {
            black = json.toJSONString()
        }
    var violationJson: JSONArray
        get() = if (violation == null) JSONArray() else JSON.parseArray(violation)
        set(json) {
            violation = json.toJSONString()
        }
    var qaJson: JSONArray
        get() = if (qa == null) JSONArray() else JSON.parseArray(qa)
        set(json) {
            qa = json.toJSONString()
        }
    var adminJson: JSONArray
        get() = if (admin == null) JSONArray() else JSON.parseArray(admin)
        set(json) {
            admin = json.toJSONString()
        }
    var weiboJson: JSONArray
        get() = if (weibo == null) JSONArray() else JSON.parseArray(weibo)
        set(json) {
            weibo = json.toJSONString()
        }
    var biliBiliJson: JSONArray
        get() = if (biliBili == null) JSONArray() else JSON.parseArray(biliBili)
        set(json) {
            biliBili = json.toJSONString()
        }
    var biliBiliLiveJson: JSONArray
        get() = if (biliBiliLive == null) JSONArray() else JSON.parseArray(biliBiliLive)
        set(json) {
            biliBiliLive = json.toJSONString()
        }
    var interceptJson: JSONArray
        get() = if (intercept == null) JSONArray() else JSON.parseArray(intercept)
        set(json) {
            intercept = json.toJSONString()
        }
    var commandLimitJson: JSONObject
        get() = if (commandLimit == null) JSONObject() else JSON.parseObject(commandLimit)
        set(json) {
            commandLimit = json.toJSONString()
        }
    var shellCommandJson: JSONArray
        get() = if (shellCommand == null) JSONArray() else JSON.parseArray(shellCommand)
        set(json) {
            shellCommand = json.toJSONString()
        }
}

interface GroupRepository: JpaRepository<GroupEntity, Int>{
    fun findByGroup(group: Long): GroupEntity?
    fun deleteByGroup(group: Long)
}

interface GroupService{
    fun findByGroup(group: Long): GroupEntity?
    fun save(groupEntity: GroupEntity): GroupEntity
    fun findAll(): List<GroupEntity>
    fun deleteByGroup(group: Long)
    fun delete(groupEntity: GroupEntity)
}

@Service
class GroupServiceImpl: GroupService{

    @Resource
    private lateinit var groupRepository: GroupRepository

    override fun findByGroup(group: Long): GroupEntity? {
        return groupRepository.findByGroup(group)
    }

    override fun save(groupEntity: GroupEntity): GroupEntity {
        return groupRepository.save(groupEntity)
    }

    override fun findAll(): List<GroupEntity> {
        return groupRepository.findAll()
    }

    override fun deleteByGroup(group: Long) {
        return groupRepository.deleteByGroup(group)
    }

    override fun delete(groupEntity: GroupEntity) {
        return groupRepository.delete(groupEntity)
    }
}